### PR TITLE
Natmgrd changes in sonic-swss sub module to support NAT feature.

### PR DESCRIPTION
**Pull request by AkhileshSamineni**
Added natmgr, and intfmgr changes.

Link to NAT HLD : https://github.com/Azure/SONiC/blob/master/doc/nat/nat_design_spec.md

Depends on:
sonic-swss : https://github.com/Azure/sonic-swss/pull/1126 and https://github.com/Azure/sonic-swss/pull/1125
sonic-swss-common :  https://github.com/Azure/sonic-swss-common/pull/304 
sonic-linux-kernel : https://github.com/Azure/sonic-linux-kernel/pull/100 
sonic-sairedis : https://github.com/Azure/sonic-sairedis/pull/519

 Signed-off-by: Akhilesh Samineni <akhilesh.samineni@broadcom.com>

### CODE REVIEW

File: cfgmgr/intfmgr.cpp [hunk 1]
```diff
@@ -152,6 +158,14 @@ bool IntfMgr::doIntfGeneralTask(const vector<string>& keys,
         // Set Interface VRF except for lo
         if (!is_lo)
         {
+            /* Set nat zone */
+            if (!nat_zone.empty())
+            {
+                FieldValueTuple t("nat_zone", nat_zone);
+                fvVector.push_back(t);
+                m_appIntfTableProducer.set(alias, fvVector);
```
Comment (marian-pritsak) [inline cfgmgr/intfmgr.cpp]: set is at the end of outer if (!is_lo)

File: cfgmgr/natmgr.cpp [hunk 2]
```diff
@@ -0,0 +1,6391 @@
...
+
+/* NatMgr Constructor */
+NatMgr::NatMgr(DBConnector *cfgDb, DBConnector *appDb, DBConnector *stateDb, const vector<string> &tableNames) :
+        Orch(cfgDb, tableNames),
+        m_cfgStaticNatTable(cfgDb, CFG_STATIC_NAT_TABLE_NAME),
+        m_cfgStaticNaptTable(cfgDb, CFG_STATIC_NAPT_TABLE_NAME),
+        m_cfgNatPoolTable(cfgDb, CFG_NAT_POOL_TABLE_NAME),
+        m_cfgNatBindingsTable(cfgDb, CFG_NAT_BINDINGS_TABLE_NAME),
+        m_cfgNatGlobalTable(cfgDb, CFG_NAT_GLOBAL_TABLE_NAME),
+        m_cfgInterfaceTable(cfgDb, CFG_INTF_TABLE_NAME),
```
Comment (marian-pritsak) [inline cfgmgr/natmgr.cpp]: intfmgr is responsible for that table

File: cfgmgr/natmgr.cpp [hunk 3]
```diff
@@ -0,0 +1,6391 @@
...
+        m_cfgLagInterfaceTable(cfgDb, CFG_LAG_INTF_TABLE_NAME),
+        m_cfgVlanInterfaceTable(cfgDb, CFG_VLAN_INTF_TABLE_NAME),
+        m_cfgLoopbackInterfaceTable(cfgDb, CFG_LOOPBACK_INTERFACE_TABLE_NAME),
+        m_cfgNatAclTable(cfgDb, CFG_ACL_TABLE_NAME),
+        m_cfgNatAclRuleTable(cfgDb, CFG_ACL_RULE_TABLE_NAME),
```
Comment (marian-pritsak) [inline cfgmgr/natmgr.cpp]: Each table has a single manager that should handle it, and these table don't belong to NAT manager.

File: cfgmgr/natmgr.cpp [hunk 4]
```diff
@@ -0,0 +1,6391 @@
...
+    
+    /* Set NAT default tcp timeout as 86400 seconds (1 Day) */
+    m_natTcpTimeout = NAT_TCP_TIMEOUT_DEFAULT;
+
+    /* Set NAT default udp timeout as 300 seconds */
+    m_natUdpTimeout = NAT_UDP_TIMEOUT_DEFAULT;
+
+    /* Clean the NAT iptables */
+    std::string res;
+    const std::string cmds = std::string("") + IPTABLES_CMD + " -F -t nat ";
```
Comment (marian-pritsak) [inline cfgmgr/natmgr.cpp]: Why clear iptables before start?

File: cfgmgr/natmgr.cpp [hunk 5]
```diff
@@ -0,0 +1,6391 @@
...
+    if (natAdminMode == ENABLED)
+    {
+        return true;
+    }
+
+    return false;
+}
+
+/* To check the give global_ip is withing the Prefix subnet or not */
+bool NatMgr::isGlobalIpMatching(const string &prefix, const string &global_ip)
```
Comment (marian-pritsak) [inline cfgmgr/natmgr.cpp]: This would fit as a method in IpPrefix class

File: cfgmgr/natmgr.cpp [hunk 6]
```diff
@@ -0,0 +1,6391 @@
...
+}
+
+/* To Add/Delete NAPT pool ip table to APPL_DB */
+void NatMgr::setNaptPoolIpTable(const string &opCmd, const string &ip_range, const string &port_range)
+{
+    uint32_t ipv4_addr_low, ipv4_addr_high, ip, setIp;
+    char ipAddr[INET_ADDRSTRLEN];
+    std::vector<swss::FieldValueTuple> values;
+
+    if (!port_range.empty() and (port_range != "NULL"))
```
Comment (marian-pritsak) [inline cfgmgr/natmgr.cpp]: First check if not NULL?

File: cfgmgr/natmgr.cpp [hunk 7]
```diff
@@ -0,0 +1,6391 @@
...
+            ipv4_addr_low = ntohl(ipv4_addr_low);
+        }
+        else
+        {
+            inet_pton(AF_INET, nat_ip[0].c_str(), &ipv4_addr_low);
+            ipv4_addr_high = ntohl(ipv4_addr_low);
+            ipv4_addr_low = ntohl(ipv4_addr_low);
+        }
+
+        for (ip = ipv4_addr_low; ip <= ipv4_addr_high; ip++)
```
Comment (marian-pritsak) [inline cfgmgr/natmgr.cpp]: Why set like this and not as a range?

File: cfgmgr/intfmgr.cpp (hunk 1 above)
Comment (AkhileshSamineni) [reply cfgmgr/intfmgr.cpp]: This set it specific to nat_zone only, so using new "fvVector" variable for it.

File: cfgmgr/natmgr.cpp (hunk 2 above)
Comment (AkhileshSamineni) [reply cfgmgr/natmgr.cpp]: I have removed these lines, but NatMgr is also a consumer for the Interface Tables, to get the configured nat_zone value on any interface, which will be used as a mark value (zone +1 ) in iptables rules.

File: cfgmgr/natmgr.cpp (hunk 3 above)
Comment (AkhileshSamineni) [reply cfgmgr/natmgr.cpp]: I have removed these lines, but NatMgr is also a consumer for the Interface Tables, to get the configured nat_zone value on every interface, which will be used as a mark value (zone +1 ) in iptables rules and ACL tables as well.

File: cfgmgr/natmgr.cpp (hunk 4 above)
Comment (AkhileshSamineni) [reply cfgmgr/natmgr.cpp]: Before the NAT application, wanted to clearing any stale nat iptables rules.

File: cfgmgr/natmgr.cpp (hunk 5 above)
Comment (AkhileshSamineni) [reply cfgmgr/natmgr.cpp]: Yes, will move it to IpPrefix class in next patch.

File: cfgmgr/natmgr.cpp (hunk 6 above)
Comment (AkhileshSamineni) [reply cfgmgr/natmgr.cpp]: No, When pool is created with only Ip address, "nat_port" value is set to "NULL" string like below

root@sonic:/home/admin# config nat add pool p1 65.55.42.5
root@sonic:/home/admin# redis-cli -n 4
127.0.0.1:6379[4]> keys *NA*
1) "NAT_POOL|p1"
127.0.0.1:6379[4]> hgetall NAT_POOL|p1
1) "nat_port"
2) "NULL"
3) "nat_ip"
4) "65.55.42.5"
127.0.0.1:6379[4]> exit
root@sonic:/home/admin#

File: cfgmgr/natmgr.cpp (hunk 7 above)
Comment (AkhileshSamineni) [reply cfgmgr/natmgr.cpp]: Here m_appNaptPoolIpTable is used by NatSyncd code to know whether the packet notified by conntrack table is NAPT'ed or NAT'ed.

For example, Create a NAT pool to do NAPT'ed like "config nat add pool p1 65.55.42.5 100-200"
If packet comes to cpu with Src Ip as 12.0.0.1 and Src port as 100, it get's SNAPT'ed to Src Ip as 65.55.42.5 and Src Port as 100, kernel picks the "100" as it is available from Pool.
So kernel doesn't set any separate flags in the conntrack entry if it is NAPT'ed. Only SNAT/DNAT flags are set in the conntrack entry that is notified to natsyncd.
When the translated port number is the same, Natsyncd will not be able to say if it is NAT'ed or NAPT'ed. So the NaptPoolIpTable is added for every IP address in the pool to help NatSyncd pick the conntrack notifications as NAPT and add them as NAPT entries in the APP_DB.

**Comment by AkhileshSamineni** [comment]
> The PR is very substantial. Can you please split it by the daemon it is related to? I think it won't break it.

**Comment by AkhileshSamineni** [comment]
> The PR is very substantial. Can you please split it by the daemon it is related to? I think it won't break it.

@marian-pritsak Yes right, Created PRs for orchanget (#1125 ) and Natsyncd (#1126 ).

**Comment by lguohan** [comment]
>             if (inet_pton(AF_INET6, ipPrefixKeys[0].c_str(), &ipv6_addr))

we have ipaddress class in sonic swss common, can you use that?

---
Refers to: cfgmgr/natmgr.cpp:6628 in d7113d6. [](commit_id = d7113d6db14a2cc1823e991f6b48426268f757e3, deletion_comment = False)

File: cfgmgr/intfmgr.cpp (hunk 1 above)
Comment (AkhileshSamineni) [reply cfgmgr/intfmgr.cpp]: After resolved merge conflicts, doing set at the end of outer if(!is_lo).

**Comment by AkhileshSamineni** [comment]
> > ```
> >         if (inet_pton(AF_INET6, ipPrefixKeys[0].c_str(), &ipv6_addr))
> > ```
> 
> we have ipaddress class in sonic swss common, can you use that?
> 
> Refers to: cfgmgr/natmgr.cpp:6628 in [d7113d6](https://github.com/Azure/sonic-swss/commit/d7113d6db14a2cc1823e991f6b48426268f757e3). [](commit_id = [d7113d6](https://github.com/Azure/sonic-swss/commit/d7113d6db14a2cc1823e991f6b48426268f757e3), deletion_comment = False)

Yes, added changes to use IpAddress class here.

**Comment by AkhileshSamineni** [comment]
> Are this pytest changes present in PR 1125. I could not fine them. Kindly confirm.

Added pytest changes in PR 1125.
Here is the link - https://github.com/Azure/sonic-swss/pull/1125/commits/69d1b68b27afceb88b94ccdc03aea1775d81a9e9

**Comment by AkhileshSamineni** [comment]
Retest this please.

**Comment by AkhileshSamineni** [comment]
retest this please.

**Comment by AkhileshSamineni** [comment]
Retest this please.

**Pull request merged**